crates/server/src/store/blocklist/authority.rs

```diff
@@ -0,0 +1,430 @@
+
+        let mut contents = String::new();
+        let _ = handle.read_to_string(&mut contents);
+
+        for mut entry in contents.split('\n') {
+            // Strip comments
+            if let Some(idx) = entry.chars().position(|c| c == '#') {
+                entry = entry[0..idx].trim();
+            }
+
+            if entry.is_empty() {
+                continue;
+            }
+
+            let mut str_entry = entry.to_string();
+            if !entry.ends_with('.') {
+                str_entry += ".";
+            }
+
+            let Ok(name) = LowerName::from_str(&str_entry[..]) else {
+                info!("unable to add blocklist entry: {str_entry}");
+                continue;
+            };
+
+            trace!("Inserting blocklist entry {str_entry}");
+
+            // The boolean value is not significant; only the key is used.
+            self.blocklist.insert(name, true);
+        }
+
+        true
+    }
+
+    /// Build a wildcard match list for a given host
+    pub fn get_wildcards(&self, host: &Name) -> Vec<LowerName> {
+        host.iter()
+            .enumerate()
+            .filter(|(i, _x)| *i > (self.min_wildcard_depth - 1) as usize)
+            .map(|(i, _x)| host.trim_to(i + 1).into_wildcard().into())
+            .collect::<Vec<LowerName>>()
+    }
+
+    /// Perform a blocklist lookup. Returns true on match, false on no match.  This is also where
+    /// wildcard expansion is done, if wildcard support is enabled for the blocklist authority.
+    fn blocklist_lookup(&self, name: &LowerName) -> bool {
+        let mut match_list = vec![name.to_owned()];
+
+        if self.wildcard_match {
+            match_list.append(&mut self.get_wildcards(&Name::from(name)));
+        }
+
+        trace!("Blocklist match list: {match_list:?}");
+
+        for host in match_list {
+            if self.blocklist.contains_key(&host) {
+                info!("block list matched query {name}");
+                return true;
+            }
+        }
+
+        false
+    }
+}
+
+#[async_trait::async_trait]
+impl Authority for BlocklistAuthority {
+    type Lookup = BlocklistLookup;
+
+    fn zone_type(&self) -> ZoneType {
+        ZoneType::Hint
+    }
+
+    fn is_axfr_allowed(&self) -> bool {
+        false
+    }
+
+    async fn update(&self, _update: &MessageRequest) -> UpdateResult<bool> {
+        Err(ResponseCode::NotImp)
+    }
+
+    fn origin(&self) -> &LowerName {
+        &self.origin
+    }
+
+    /// Perform a blocklist lookup.  This will return LookupControlFlow::Break(Ok) on a match, or
+    /// LookupControlFlow::Skip on no match.
+    async fn lookup(
+        &self,
+        name: &LowerName,
+        rtype: RecordType,
+        _lookup_options: LookupOptions,
+    ) -> LookupControlFlow<Self::Lookup> {
+        use LookupControlFlow::*;
+
+        trace!("blocklist lookup: {name} {rtype}");
+
+        if self.blocklist_lookup(name) {
+            return Break(Ok(BlocklistLookup(Lookup::from_rdata(
```

Comment:
I could add an option for including a TXT record.  I've never seen that implemented before in a DNS sinkhole - usually the goal is to drop the connections silently (ad blocking use cases) or direct clients to a honeypot or sinkhole (malware use cases,) but I could see it being useful is some situations.

Reply:
Those records would effectively be ignored, and I think wouldn't change the behavior? In most cases I think the client would never see them because the resolution would ignore the NS or TXT records in the Response. I was thinking of it mainly for debugging purposes I guess.

Reply:
There's an option now to set 'block_message' in the block list config; if that is set, a TXT record will be returned along with the A/AAAA record.